Signal/src/ViewControllers/ConversationView/MessagesViewController.m

```diff
@@ -3521,9 +3521,27 @@ - (BOOL)isScrolledToBottom
         return NO;
     }
 
+    // This is a bit subtle.
+    //
+    // The _wrong_ way to determine if we're scrolled to the bottom is to
+    // measure whether the collection view's content is "near" the bottom edge
+    // of the collection view.  This is wrong because the collection view
+    // might not have enough content to fill the collection view's bounds
+    // _under certain conditions_ (e.g. with the keyboard dismissed).
+    //
+    // What we're really interested in is something a bit more subtle:
+    // "Is the scroll view scrolled down as far as it can, "at rest".
+    //
+    // To determine that, we find the appropriate "content offset y" if
+    // the scroll view were scrolled down as far as possible.  IFF the
+    // actual "content offset y" is "near" that value, we return YES.
     const CGFloat kIsAtBottomTolerancePts = 5;
-    return (self.collectionView.contentOffset.y + self.collectionView.bounds.size.height + kIsAtBottomTolerancePts
-        >= self.collectionView.contentSize.height);
+    // Note the usage of MAX() to handle the case where there isn't enough
+    // content to fill the collection view at its current size.
+    CGFloat contentOffsetYBottom
```

Comment:
Would it be equally correct to leave the existing code, but add a guard like this before it?

```
if (self.collectionView.contentSize.height <= self.collectionView.bounds.size.height + kIsAtBottomTolerancePts) {
    return YES;
}
```

IMO it'd be a little easier to follow, and might obviate the need for the verbose comment.

Reply:
I think so but it will be significantly harder to test this condition, since its easy to change scroll state in a fine-grained way but incredibly difficult to manipulate content size in the same way.  I think I'll leave this as-is.